examples/src/bin/gio_task.rs

```diff
@@ -0,0 +1,263 @@
+                .await
+                .unwrap()
+                .get_size();
+
+            let source_object = task
+                .upcast_ref::<gio::AsyncResult>()
+                .get_source_object()
+                .unwrap();
+
+            let source_object =
+                imp::FileSize::from_instance(&source_object.downcast_ref::<FileSize>().unwrap());
+
+            source_object.size.replace(Some(size));
+            task.return_value(&size.to_value());
+        });
+    }
+}
+
+impl Default for FileSize {
+    fn default() -> Self {
+        FileSize::new()
+    }
+}
+
+// This function mimicks what the C code using the exported async/finish API would do. It first defines a
+// callback of type GAsyncResult that internally calls get_file_size_finish to retrieve the Task
+// result, and simply prints it out. Then it invokes the my_file_size_get_file_size_async method, passing the
+// callback to it as parameter.
+async fn run_unsafe() {
+    let simple_object = FileSize::new();
+    let cancellable = gio::Cancellable::new();
+
+    // The callback to be passed to my_file_size_get_file_size_async
+    unsafe extern "C" fn c_callback(
+        source_object: *mut glib::object::GObject,
+        result: *mut gio::ffi::GAsyncResult,
+        _user_data: glib::ffi::gpointer,
+    ) {
+        let mut error = std::ptr::null_mut();
+        let ret = ffi::my_file_size_get_file_size_finish(
+            source_object as *mut ffi::FileSize,
+            result,
+            &mut error,
+        );
+        if !error.is_null() {
+            eprintln!("Task returned error!");
+            return;
+        }
+
+        println!("Unsafe callback - Returned value from task: {}", ret);
+        println!(
+            "Unsafe callback - FileSize::size: {}",
+            ffi::my_file_size_get_retrieved_size(source_object as *mut ffi::FileSize)
+        );
+    }
+
+    // The actual call to my_file_size_get_file_size_async
+    unsafe {
+        ffi::my_file_size_get_file_size_async(
+            simple_object.to_glib_none().0,
+            cancellable.to_glib_none().0,
+            Some(c_callback),
+            std::ptr::null_mut(),
+        );
+    }
+}
+
+// This function is the "safe" counterpart of "run_unsafe", using only safe rust
+// bindings to accomplish the same task
+async fn run_safe() {
+    let simple_object = FileSize::new();
+    let cancellable = gio::Cancellable::new();
+
+    fn callback(value: i64, source_object: &FileSize) {
+        println!("Safe callback - Returned value from task: {}", value);
+        println!(
+            "Safe callback - FileSize::size: {}",
+            source_object.retrieved_size()
+        );
+    }
+
+    simple_object.get_file_size_async(Some(&cancellable), callback);
+}
+
+fn main() {
+    let c = glib::MainContext::default();
+    let l = glib::MainLoop::new(Some(&c), false);
+
+    c.push_thread_default();
+    c.spawn_local(run_unsafe());
+    c.spawn_local(run_safe());
+    l.run();
```

Comment:
Would be nice to make it so that the mainloop quits once both callbacks were called